package-res/resources/web/pentaho/lang/SortedList.js

```diff
@@ -0,0 +1,228 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+define([
+  "./List",
+  "./Base"
+], function(List, Base) {
+
+  var baseProto = Base.Array.prototype;
+
+  return List.extend("pentaho.lang.SortedList", /** @lends  pentaho.lang.SortedList# */{
+    /**
+     * @classdesc The `SortedList` class is an abstract base class for ordered arrays.
+     *
+     * ### AMD
+     *
+     * To obtain the constructor of this class,
+     * require the module `"pentaho/lang/SortedList"`.
+     *
+     * ### Remarks
+     *
+     * If an attempt is made to add a element to a specific index of the list
+     * an error is thrown.
+     *
+     * @class
+     * @name SortedList
+     * @memberOf pentaho.lang
+     * @abstract
+     * @extends pentaho.lang.List
+     *
+     * @description Initializes a sorted list instance.
+     *
+     * Note that because a `SortedList` is a sub-class of `Array`,
+     * it cannot be a "newable" constructor function (at least up to ECMAScript version 5).
+     * Instead, instances of `SortedList` are actually initial instances of `Array`
+     * whose prototype is then changed to be that of `SortedList`.
+     *
+     * In other words, `SortedList` is an "initialization" constructor
+     * (see {@link pentaho.lang.ISpecifiable} for more information
+     *  on these concepts).
+     *
+     * Concrete `SortedList` sub-classes should provide a static `to` method
+     * to help in their construction.
+     *
+     * @param {Object}   [keyArgs]         The keyword arguments.
+     * @param {Function} [compareFunction] Specifies a function that defines the sort order.
+     *
+     * These are not used directly by the `SortedList` class
+     * but are passed-through to the methods that handle
+     * the initialization of each list element.
+     */
+    constructor: function(keyArgs, compareFunction) {
+      if (compareFunction) {
+        this.compare = compareFunction;
+      }
+
+      this.base(keyArgs);
+    },
+
+    _compareFunction: function(e1, e2) {
+      if (e1 == null || e2 == null) {
+        if (e1 != null) {
+          return 1;
+        }
+
+        if (e2 != null) {
+          return -1;
+        }
+
+        return 0;
+      }
+
+      var v1 = e1.valueOf();
+      var v2 = e2.valueOf();
+
+      return v1 === v2 ? 0 : (v1 > v2 ? 1 : -1);
+    },
+
+    get compare() {
+      return this._compareFunction;
+    },
+
+    set compare(compareFunction) {
+      this._compareFunction = compareFunction;
+    },
+
+    sort: function(compareFunction) {
+      baseProto.sort.call(this, compareFunction || this._compareFunction);
+    },
+
+    /**
+     * Performs a binary search on the sorted list.
+     *
+     * @param {*} elem The item to search for.
+     *
+     * @return {Number} The index of the element;
```

Comment:
The ~ complement operation (as a side effect, I admit) when dealing with integers results in `~x = (-1 * x) - 1`.

So lets recapitulate:

* When not found, we get the _negative index minus 1_ of the _last smaller element_. `Math.abs` it and we get the index to insert our element **after** that one.
 * In an empty list, or in a list with only _bigger elements_, it will be equal to `0`. In a list with only _smaller elements_ it will be equal to `.length`. So it makes sense.
* When found we get the index of the first _equal element_. Inserting at it adds the new element **before** that one.

Maybe this second case is the one that can prevent
> Within a type configuration, if all other rule ordering criteria are equal, the later configuration rules override the former configuration rules.